### PR TITLE
Amor detailed data reduction document

### PR DESCRIPTION
**Pull request by arm61**
This is a first draft of the detailed data reduction documentation for the Amor instrument. The purpose of this notebook will be to share the reduction methodology in an understandable and open fashion.

### CODE REVIEW

**Comment by arm61** [comment]
This is the first draft of this document. My plan is to get this draft on the `ess-notebooks` webpage and then share with ESS and PSI colleagues for revision.

But first! `scipp` review would be massively appreciated 🙂

File: reflectometry/amor_reduction_detailed.ipynb [hunk 1]
```diff
@@ -0,0 +1,1106 @@
...
+   "id": "ignored-chemical",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "chopper_sample_distance = 15.0 * sc.units.m\n",
+    "data.attrs[\"source_position\"] = sc.geometry.position(\n",
+    "    0.0 * sc.units.m, \n",
+    "    0.0 * sc.units.m, \n",
+    "    -chopper_sample_distance)"
+   ]
```
Comment (SimonHeybrock) [inline reflectometry/amor_reduction_detailed.ipynb:300]: Use the new `sc.vector()` which is briefer and more understandable?

File: reflectometry/amor_reduction_detailed.ipynb [hunk 2]
```diff
@@ -0,0 +1,1106 @@
...
+   "metadata": {
+    "scrolled": true
+   },
+   "outputs": [],
+   "source": [
+    "max_w = sc.max(data.events.coords['wavelength'])\n",
+    "min_w = sc.min(data.events.coords['wavelength'])\n",
+    "bins = sc.Variable(values=np.linspace(min_w.value, max_w.value, 100), \n",
+    "                   unit=data.events.coords['wavelength'].unit,\n",
+    "                   dims=['wavelength'])\n",
```
Comment (SimonHeybrock) [inline reflectometry/amor_reduction_detailed.ipynb:353]: `sc.linspace`?

File: reflectometry/amor_reduction_detailed.ipynb [hunk 3]
```diff
@@ -0,0 +1,1106 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "elder-filling",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "data.events"
```
Comment (SimonHeybrock) [inline reflectometry/amor_reduction_detailed.ipynb:621]: For docs, showing underlying events as table may be useful, e.g.:

```python
sc.table(data.events['event', :5])
```

File: reflectometry/amor_reduction_detailed.ipynb [hunk 4]
```diff
@@ -0,0 +1,1106 @@
...
+   "outputs": [],
+   "source": [
+    "beam_size = 0.001 * sc.units.m\n",
+    "sample_size = 0.5 * sc.units.m\n",
+    "beam_on_sample = beam_size / sc.sin(data.bins.coords['theta'])\n",
+    "half_beam_on_sample = (beam_on_sample / 2.0)\n",
+    "offset_positive = sc.geometry.position(\n",
+    "    data.coords[\"sample_position\"].fields.x, \n",
+    "    data.coords[\"sample_position\"].fields.y, \n",
+    "    data.coords[\"sample_position\"].fields.z + half_beam_on_sample)\n",
```
Comment (SimonHeybrock) [inline reflectometry/amor_reduction_detailed.ipynb:835]:
```python
offset_positive = da.coords['sample_position'].copy()
offset_positive.fields.z + half_beam_on_sample
```
?

File: reflectometry/amor_reduction_detailed.ipynb [hunk 5]
```diff
@@ -0,0 +1,1106 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "id": "f317df7a",
+   "metadata": {},
+   "source": [
+    "These three contributors to the resolution function are combined to give the total resolution function, \n",
+    "\n",
+    "$$\n",
+    "\\frac{\\sigma q_z}{q_z} = \\sqrt{\\bigg(\\frac{\\sigma\\lambda}{\\lambda}\\bigg)^2+\\bigg(\\frac{\\sigma\\gamma}{\\theta}\\bigg)^2+\\bigg(\\frac{\\sigma\\theta}{\\theta}\\bigg)^2}\n",
```
Comment (SimonHeybrock) [inline reflectometry/amor_reduction_detailed.ipynb:867]: Looks like the norm of a vector, but I suppose storing the three components as a single vector is not quite working for you?

**Review by SimonHeybrock — COMMENTED** [review]
Didn't check details and equations, a couple of comments, looks like a good start otherwise:

File: reflectometry/amor_reduction_detailed.ipynb [hunk 6]
```diff
@@ -0,0 +1,1106 @@
...
+    "data"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "owned-button",
+   "metadata": {},
+   "source": [
+    "The data is stored in a binned structure, where each bin is a pixel on the detector containing each neutron event that was measured there. \n",
+    "It is possible to access the event data from the class property `events`. "
```
[the remark this answers is not in the document]
Comment (arm61) [reply reflectometry/amor_reduction_detailed.ipynb:91]: Screenshot 2021-07-08 at 08 49 36
`sc.show(data)` doesn't look so great for real data.

[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply reflectometry/amor_reduction_detailed.ipynb:91]: Can you try `sc.show(sc.transpose(data))`?

[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply reflectometry/amor_reduction_detailed.ipynb:91]: But yes, this is a known issue, but so far we have not had a good idea on how to fix this (without excessive development effort). :(

File: reflectometry/amor_reduction_detailed.ipynb (hunk 5 above)
Comment (arm61) [reply reflectometry/amor_reduction_detailed.ipynb:867]: It is just incidental that there are three components, these three are the largest contributors to the resolution function but in theory there could be many more. I think that using a vector would make it more confusing for the reader.

**Comment by arm61** [comment]
> Didn't check details and equations, a couple of comments, looks like a good start otherwise:

Once you are happy that the code is as good as it can be we can merge and share with the others. This is of course going to go through multiple iterations I expect.

File: reflectometry/amor_reduction_detailed.ipynb [hunk 7]
```diff
@@ -348,12 +349,35 @@
...
+    "    unit=data.events.coords['wavelength'].unit)\n",
     "sc.histogram(data.events, bins).plot()"
    ]
   },
+  {
+   "cell_type": "markdown",
+   "id": "7fb6a9fc",
+   "metadata": {},
+   "source": [
+    "In addition to plotting a histogram of the events as a function of wavelength, we can visualise a subset in a table."
```
Comment (SimonHeybrock) [inline reflectometry/amor_reduction_detailed.ipynb:366]: Actually you can visualize all: You get a slider if you omit the `['event', :10]`.

**Review by SimonHeybrock — APPROVED** [review]

File: reflectometry/amor_reduction_detailed.ipynb (hunk 7 above)
Comment (arm61) [reply reflectometry/amor_reduction_detailed.ipynb:366]: Is it possible to have the number of rows spinbox default to a smaller value than 50? Something like 

`sc.table(data['event', :], nrows=10)`

File: reflectometry/amor_reduction_detailed.ipynb (hunk 6 above)
[the remark this answers is not in the document]
Comment (arm61) [reply reflectometry/amor_reduction_detailed.ipynb:91]: `sc.transpose(data, dims=['detector_id'])`? this gives an error so I must have the transposition dimension wrong

[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply reflectometry/amor_reduction_detailed.ipynb:91]: Oh, I guess your data is just 1D then, not (1xN)?

[the remark this answers is not in the document]
Comment (arm61) [reply reflectometry/amor_reduction_detailed.ipynb:91]: Yeah I think so.

[the remark this answers is not in the document]
Comment (arm61) [reply reflectometry/amor_reduction_detailed.ipynb:91]: That makes sense from the image above actually.

File: reflectometry/amor_reduction_detailed.ipynb (hunk 7 above)
Comment (nvaytet) [reply reflectometry/amor_reduction_detailed.ipynb:366]: it is actually using a setting from the config:
https://github.com/scipp/scipp/blob/main/python/src/scipp/table.py#L292
but it's true that we could make it a function argument.

**Comment by arm61** [comment]
I reckon we can merge this as is then, once something happens with https://github.com/scipp/scipp/issues/2034 I will update (I don't have merge power here)

File: reflectometry/amor_reduction_detailed.ipynb (hunk 7 above)
Comment (SimonHeybrock) [reply reflectometry/amor_reduction_detailed.ipynb:366]: See https://github.com/scipp/scipp/issues/2034.

File: reflectometry/detailed_plots.py [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+
+    def dtrace(z, z0):
+        return ((-a * (z1 - z0) ** 2) / (2 * v ** 2) - y0 + y1) * 1 / (z1 - z0) + a * (z - z0) / (v ** 2)
+
+    z = np.linspace(0.00001, z1-0.01, 1000)
+    z_diff = np.linspace(0.4, z1-0.01, 1000)
+    y = np.linspace(0.00001, y1-0.001, 1000)
+
+    fig, ax = plt.subplots(figsize=(8, 4))
+    plt.plot(z, trace(z, 0))
```
Comment (nvaytet) [inline reflectometry/detailed_plots.py:76]: I really recommend here to use `ax.plot` instead of `plt.plot`

File: reflectometry/detailed_plots.py [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+    plt.plot(z, z * dtrace(0, 0))
+    plt.plot(z, np.linspace(0.00001, trace(z, 0)[-1], 1000), ls='--')
+    plt.xlim(-1, 4.8)
+    plt.ylim(-0.3, 1.5)
+    rectangle = plt.Rectangle((-0.75,-0.25), 1.5, 0.25, ec='k', fc=(1, 0, 0, 0))
+    plt.gca().add_patch(rectangle)
```
Comment (nvaytet) [inline reflectometry/detailed_plots.py:82]:
```suggestion
    ax.add_patch(rectangle)
```

File: reflectometry/detailed_plots.py [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+
+    z = np.linspace(0.00001, z1-0.01, 1000)
+    z_diff = np.linspace(0.4, z1-0.01, 1000)
+    y = np.linspace(0.00001, y1-0.001, 1000)
+
+    fig, ax = plt.subplots(figsize=(8, 4))
+    plt.plot(z, trace(z, 0))
+    plt.plot(z, z * dtrace(0, 0))
+    plt.plot(z, np.linspace(0.00001, trace(z, 0)[-1], 1000), ls='--')
+    plt.xlim(-1, 4.8)
```
Comment (nvaytet) [inline reflectometry/detailed_plots.py:79]:
```suggestion
    ax.set_xlim(-1, 4.8)
```

File: reflectometry/detailed_plots.py [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+    plt.ylim(-0.3, 1.5)
+    rectangle = plt.Rectangle((-0.75,-0.25), 1.5, 0.25, ec='k', fc=(1, 0, 0, 0))
+    plt.gca().add_patch(rectangle)
+    rectangle = plt.Rectangle((z1,0.3), 0.4, 1, ec='k', fc=(1, 0, 0, 0))
+    plt.text(0, -0.125, 'Sample', horizontalalignment='center', verticalalignment='center')
+    plt.text(z1+0.2, 0.8, 'Detector', horizontalalignment='center', verticalalignment='center', rotation='vertical')
+    plt.gca().add_patch(rectangle)
+    plt.xlabel('$z$/m')
+    plt.ylabel('$y$/m')
+    plt.savefig('gravity.png')
```
Comment (nvaytet) [inline reflectometry/detailed_plots.py:89]:
```suggestion
    fig.savefig('gravity.png')
```

File: reflectometry/detailed_plots.py [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+    plt.close()
```
Comment (nvaytet) [inline reflectometry/detailed_plots.py:90]:
```suggestion
    plt.close(fig)
```

**Review by nvaytet — CHANGES_REQUESTED** [review]

File: reflectometry/detailed_plots.py (hunk 8 above)
Comment (arm61) [reply reflectometry/detailed_plots.py:76]: Sob, I like my mess of coding paradigms in `mpl` though 🤣. Sorting now

**Review by nvaytet — APPROVED** [review]

**Comment by arm61** [comment]
@nvaytet or @SimonHeybrock can you merge this now?

**Pull request merged**